File: Editorial/107_Binary_Tree_Level_Order_Traversal_II/C/1_Recursion_DFS_Preorder_Traversal.c
void helper(struct TreeNode* node, int level, int* returnSize,
            int** columnSizes, int*** levels) {
    // start the current level
    if (*returnSize == level) {
        (*returnSize)++;
        *columnSizes = (int*)realloc(*columnSizes, (*returnSize) * sizeof(int));
        (*columnSizes)[level] = 0;
        *levels = (int**)realloc(*levels, (*returnSize) * sizeof(int*));
        (*levels)[level] = (int*)malloc(1000 * sizeof(int));
    }
    // append the current node value
    (*levels)[level][(*columnSizes)[level]++] = node->val;
    // process child nodes for the next level
    if (node->left != NULL)
        helper(node->left, level + 1, returnSize, columnSizes, levels);
    if (node->right != NULL)
        helper(node->right, level + 1, returnSize, columnSizes, levels);
}
int** levelOrderBottom(struct TreeNode* root, int* returnSize,
                       int** columnSizes) {
    *returnSize = 0;
    *columnSizes = NULL;
    int** levels = NULL;
    if (root != NULL) helper(root, 0, returnSize, columnSizes, &levels);
    // reverse the levels
    for (int i = 0; i < *returnSize / 2; i++) {
        int* tmp = levels[i];
        levels[i] = levels[*returnSize - 1 - i];
        levels[*returnSize - 1 - i] = tmp;
        int tmp_size = (*columnSizes)[i];
        (*columnSizes)[i] = (*columnSizes)[*returnSize - 1 - i];
        (*columnSizes)[*returnSize - 1 - i] = tmp_size;
    }
    return levels;
}

File: Editorial/0101-0200/0107_Binary_Tree_Level_Order_Traversal_II/CSharp/1_Recursion_DFS_Preorder_Traversal.cs
public class Solution {
    List<IList<int>> levels = new List<IList<int>>();

    public void Helper(TreeNode node, int level) {
        if (levels.Count == level)
            levels.Add(new List<int>());
        levels[level].Add(node.val);
        if (node.left != null)
            Helper(node.left, level + 1);
        if (node.right != null)
            Helper(node.right, level + 1);
    }

    public IList<IList<int>> LevelOrderBottom(TreeNode root) {
        if (root == null)
            return levels;
        Helper(root, 0);
        levels.Reverse();
        return levels;
    }
}
